common/lib/xmodule/xmodule/modulestore/__init__.py

```diff
@@ -139,6 +140,20 @@ def unnest(self):
         """
         self._active_count -= 1
 
+    def add_publish_item(self, location):
+        """
+        Adds an item to the set of things that have been published
+        """
+        self._publish_items.add(location.version_agnostic().for_branch(None))
```

Comment:
If the choice is between letting it grow without bound and setting an arbitrary cap, I'd take the latter. This sounds like it could get complicated though.

Stepping back just a bit, is the goal just to make sure we don't re-index things that are already covered by the index? If so, could you just check for `subtree_edited_on` on the descriptors as you're crawling down through the course, and only re-index things that are past a certain date? You'd have to add the `EditInfoMixin` to the list of XBLOCK_MIXINS on the LMS (you can lift that straight out of my grades performance PR https://github.com/edx/edx-platform/pull/7288/files#diff-82bda515879c96ab0641bfcbe58631c3R629).

What is the performance penalty for doing a full course re-index, anyhow?